examples/WGAN-GP/model.py

```diff
@@ -0,0 +1,142 @@
+            padding = "VALID" if i == 0 else "SAME"
+            x = elegy.nn.conv.Conv2DTranspose(
+                c, (4, 4), stride=(2, 2), padding=padding
+            )(x)
+            x = elegy.nn.BatchNormalization(decay_rate=0.9)(x)
+            x = jax.nn.leaky_relu(x, negative_slope=0.2)
+        x = elegy.nn.conv.Conv2DTranspose(3, (4, 4), stride=(2, 2))(x)
+        x = jax.nn.sigmoid(x)
+        return x
+
+
+# the discriminator architecture adapted from DCGAN
+# also called 'critic' in the WGAN paper
+class Discriminator(elegy.Module):
+    def call(self, x):
+        for c in [128, 256, 512, 1024]:
+            x = elegy.nn.conv.Conv2D(c, (4, 4), stride=(2, 2))(x)
+            x = jax.nn.leaky_relu(x, negative_slope=0.2)
+        x = elegy.nn.Flatten()(x)
+        x = elegy.nn.Linear(1)(x)
+        return x
+
+
+# multiplier for gradient normalization
+LAMBDA_GP = 10
+
+# gradient regularization term
+def gradient_penalty(x_real, x_fake, applied_discriminator_fn, rngkey):
+    assert len(x_real) == len(x_fake)
+    alpha = jax.random.uniform(rngkey, shape=[len(x_real), 1, 1, 1])
+    x_hat = x_real * alpha + x_fake * (1 - alpha)
+    grads = jax.grad(lambda x: applied_discriminator_fn(x)[0].mean())(x_hat)
+    norm = jnp.sqrt((grads ** 2).sum(axis=[1, 2, 3]))
+    penalty = (norm - 1) ** 2
+    return penalty.mean() * LAMBDA_GP
+
+
+class WGAN_GP(elegy.Model):
+    def __init__(self):
+        super().__init__()
+        self.generator = Generator()
+        self.discriminator = Discriminator()
+        self.g_optimizer = optax.adam(2e-4, b1=0.5)
+        self.d_optimizer = optax.adam(2e-4, b1=0.5)
+
+    def init(self, x):
+        rng = elegy.RNGSeq(0)
+        gx, g_params, g_states = self.generator.init(rng=rng)(x)
+        dx, d_params, d_states = self.discriminator.init(rng=rng)(gx)
+
+        g_optimizer_states = self.g_optimizer.init(g_params)
+        d_optimizer_states = self.d_optimizer.init(d_params)
+
+        self.states = elegy.States(
+            g_states=g_states,
+            d_states=d_states,
+            g_params=g_params,
+            d_params=d_params,
+            g_opt_states=g_optimizer_states,
+            d_opt_states=d_optimizer_states,
+            rng=rng,
+            step=0,
+        )
+        self.initial_states = self.states.copy()
+
+    def pred_step(self, x, states):
+        z = x
+        x_fake = self.generator.apply(states.g_params, states.g_states)(z)[0]
+        return (x_fake, states)
+
+    def train_step(self, x, states):
+        # training the discriminator on every iteration
+        d_loss, gp, states = self.discriminator_step(x, states)
+
+        # training the generator only every 5 iterations as recommended in the original WGAN paper
+        step = states.step + 1
+        no_update = lambda args: (0.0, args[1])
+        do_update = lambda args: self.generator_step(len(args[0]), args[1])
+        g_loss, states = jax.lax.cond(step % 5 == 0, do_update, no_update, (x, states))
```

Comment:
I believe this should work and its a bit more readable:

```python
g_loss, states = jax.lax.cond(
    step % 5 == 0, 
    lambda _: (0.0, states), 
    lambda _: self.generator_step(len(x), states), 
    None,
)
```
Not sure when exactly the `operand` argument is necessary.

Reply:
No this doesn't work, the operand is required:
```
jax.core.UnexpectedTracerError: Encountered an unexpected tracer. Perhaps this tracer escaped throug
h global state from a previously traced function.                                                   
The functions being transformed should not save traced values to global state.                      
The tracer that caused this error was created on line elegy/elegy/types.py:66 (next).                                                          
When the tracer was created, the final 5 stack frames (most recent last) excluding JAX-internal frames were:
elegy/elegy/utils.py:101 (wrapper)
elegy/examples/WGAN-GP/model.py:89 (train_step)
elegy/examples/WGAN-GP/model.py:88 (<lambda>)
elegy/examples/WGAN-GP/model.py:125 (generator_step)
elegy/elegy/types.py:66 (next)
The function being traced when the tracer leaked was <lambda> at elegy/examples/WGAN-GP/model.py:88.
To catch the leak earlier, try setting the environment variable JAX_CHECK_TRACER_LEAKS or using the
`jax.checking_leaks` context manager.

```

Reply:
Changed to a mix which I for one find more readable:

```python
no_update = lambda states: (0.0, states)
do_update = lambda states: self.generator_step(len(x), states)
g_loss, states = jax.lax.cond(step % 5 == 0, do_update, no_update, states)
```

Reply:
Interesting, I got it working with a simple example but maybe it was too simple. This looks good!